apps/aadl-eventdata-direct-new/components/Sender/src/sender.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2017, Data61
+ * Commonwealth Scientific and Industrial Research Organisation (CSIRO)
+ * ABN 41 687 119 230.
+ *
+ * Copyright 2019 Adventium Labs
+ * Modifications made to original
+ *
+ * This software may be distributed and modified according to the terms of
+ * the BSD 2-Clause license. Note that NO WARRANTY is provided.
+ * See "LICENSE_BSD2.txt" for details.
+ *
+ * @TAG(DATA61_Adventium_BSD)
+ */
+
+#include <camkes.h>
+#include <stdio.h>
+#include <string.h>
+#include <sel4/sel4.h>
+#include <stdlib.h>
+#include <counter.h>
+#include <data.h>
+#include <queue.h>
+
+//------------------------------------------------------------------------------
+// Implementation of AADL Input Event Data Port (out) named "p1_out"
+//
+// NOTE: If we only need polling style receivers, we can get rid of the SendEvent
+
+// Assumption: only one thread is calling this and/or reading p1_in_recv_counter.
+void p1_out_aadl_event_data_send(data_t *data) {
+    queue_enqueue(p1_out_queue, data);
+    p1_out_SendEvent_emit();
+}
+
+//------------------------------------------------------------------------------
+// Testing
+
+int run(void) {
+
+    int i = 0;
+    int err = 0;
+    data_t data;
+
+    queue_init(p1_out_queue);
```

Comment:
Could we use CAmkES init functions to do this?  (see pre_init and post_init in https://github.com/seL4/camkes-tool/blob/master/docs/index.md)

Reply:
Great suggestion.  That fixes a concern I had and forgotten about. It makes sure queue init happens before sender and receivers start. I added commit making the change to the pull request.